jax/lax/lax.py

```diff
@@ -2558,18 +2558,29 @@ def _reshape_impl(operand, new_sizes, dimensions, old_sizes):
       aval = ShapedArray(new_sizes, operand.dtype)
       lazy_expr = lazy.broadcast(operand._lazy_expr, new_sizes, bcast_dims)
       return xla.DeviceArray(aval, None, lazy_expr, operand.device_buffer)
-  if (type(operand) is pxla.ShardedDeviceArray and dimensions is None
-      and _is_axis_merge(old_sizes, new_sizes)):
-    aval = ShapedArray(new_sizes, operand.dtype)
-    return pxla.ChunkedDeviceArray(old_sizes[0], aval, operand.device_buffers)
-  elif (type(operand) is pxla.ChunkedDeviceArray and dimensions is None
-        and _is_axis_split(old_sizes, new_sizes)
-        and operand.axis_size == new_sizes[0]):
-    aval = ShapedArray(new_sizes, operand.dtype)
-    return pxla.ShardedDeviceArray(aval, operand.device_buffers)
-  else:
-    return xla.apply_primitive(reshape_p, operand, new_sizes=new_sizes,
-                               dimensions=dimensions, old_sizes=old_sizes)
+
+  if type(operand) is pxla.ShardedDeviceArray and dimensions is None:
```

Comment:
I think you're right! I think this works for now, because it assumes all SDAs are only sharded across leading axes, and our current APIs only produce SDAs with this property. However, we'll have to update this once this isn't the case. I'll add a TODO -- I think once we develop more infra for dealing with indices (e.g. to support efficient nested pmaps) this will be easier to address.